### PR TITLE
Faster ID generation

### PR DESCRIPTION
**Pull request by lbloder**
## :scroll: Description
As proposed in #3809, this PR implements a faster way to generate IDs used in Sentry.
In addition to the proposal two methods were added to directly generate an ID String without the need to call two methods:
- `SentryUUID.generateSentryId()`
- `SentryUUID.generateSpanId()`

`SentryUUID.generateSpanId()` only generates the first part of a UUID as before we would create a full uuid and then only use the first 16 out of 32 characters. Therefore we only need to create 8 random bytes instead of 16.

## :bulb: Motivation and Context
Implements #3809

## :pencil: Checklist

- [ ] I reviewed the submitted code.
- [ ] I added tests to verify the changes.
- [ ] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [ ] I updated the docs if needed.
- [ ] Review from the native team if needed.
- [ ] No breaking change or entry added to the changelog.
- [ ] No breaking change for hybrid SDKs or communicated to hybrid SDKs.

### CODE REVIEW

File: sentry/src/main/java/io/sentry/SentryUUID.java [hunk 1]
```diff
@@ -0,0 +1,299 @@
+package io.sentry;
+
+import java.util.Arrays;
+import java.util.Random;
+import java.util.UUID;
+
+/**
+ * Utility class for faster id generation for SentryId, SpanId, and unique filenames uses throughout
+ * the SDK It uses our vendored Random class instead of SecureRandom for improved performance It
```
Comment (romtsn) [inline sentry/src/main/java/io/sentry/SentryUUID.java:9]: But it doesn't 😅

File: sentry/src/main/java/io/sentry/SentryUUID.java [hunk 2]
```diff
@@ -0,0 +1,299 @@
+package io.sentry;
+
+import java.util.Arrays;
+import java.util.Random;
```
Comment (romtsn) [inline sentry/src/main/java/io/sentry/SentryUUID.java:4]:
```suggestion
import io.sentry.util.Random;
```

File: sentry/src/main/java/io/sentry/SentryUUID.java (hunk 1 above)
Comment (lbloder) [reply sentry/src/main/java/io/sentry/SentryUUID.java:9]: not yet :)
Sorry, wanted to add a comment here in the PR, that we need to first merge the `main` branch to 8.x.x and then to this one

File: sentry/src/main/java/io/sentry/SentryUUID.java [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+  /**
+   * Parses a UUID from the given character sequence. The character sequence must represent a UUID
+   * as described in {@link UUID#toString()}.
+   *
+   * @param uuidSequence the character sequence from which to parse a UUID
+   * @return the UUID represented by the given character sequence
+   * @throws IllegalArgumentException if the given character sequence does not conform to the string
+   *     representation as described in {@link UUID#toString()}
+   */
+  public static UUID parseUUID(final CharSequence uuidSequence) {
```
Comment (lbloder) [inline sentry/src/main/java/io/sentry/SentryUUID.java:70]: Not sure if we even need this

File: sentry/src/main/java/io/sentry/SentryUUID.java [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+  }
+
+  public static UUID randomUUID() {
+    Random random = SentryUUID.Holder.numberGenerator;
+    byte[] randomBytes = new byte[16];
+    random.nextBytes(randomBytes);
+    randomBytes[6] &= 0x0f; /* clear version        */
+    randomBytes[6] |= 0x40; /* set to version 4     */
+    randomBytes[8] &= 0x3f; /* clear variant        */
+    randomBytes[8] |= (byte) 0x80; /* set to IETF variant  */
```
Comment (lbloder) [inline sentry/src/main/java/io/sentry/SentryUUID.java:282]: This method is taken from `java.uitl.UUID`.
Do we need this part? Do we want/need to add those flags to our random ids?

**Comment by lbloder** [comment]
In addition to the proposal in #3809, which focused on the generation of a 32 char `SentryId`, this PR also aims to further improve the generation of `SpanId` which is only 16 char in length.
Previously a SpanId was generated like this `UUID.randomUUID().toString().replace("-", "").substring(0, 16)`.

Therefore there is even more room for optimization here, as we do not actually need to generate a full UUID, e.g. 16 random bytes, but only 8 random bytes as only half of the resulting String is being used. Also, when generating only the needed 8 bytes the `substring` part can also be eliminated.

I did some benchmarks on this with the same parameters as the benchmarks in  #3809 comparing the original way of generating a SpanId with the changes of this PR:

## Java 17
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/lukasbloder/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/lukasbloder/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">

</head>

<body link="#467886" vlink="#96607D">


Benchmark | Score |   | Error |   |  
-- | -- | -- | -- | -- | --
generateSpanIdSecureRandomToStringReplaceAndSubstring16 | 2357463,435 | ± | 11044,652 | ops/s | 100%
generateSpanIdRandomFastUUIDString | 28488104,485 | ± | 236823,660 | ops/s | 1208%



</body>

</html>


## Java 8

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/lukasbloder/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/lukasbloder/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">

</head>

<body link="#467886" vlink="#96607D">


Benchmark | Score |   | Error |   |  
-- | -- | -- | -- | -- | --
generateSpanIdSecureRandomToStringReplaceAndSubstring16 | 1017070,653 | ± | 5762,135 | ops/s | 100%
generateSpanIdRandomFastUUIDString | 28461132,111 | ± | 122920,961 | ops/s | 2798%



</body>

</html>

File: sentry/src/main/java/io/sentry/SentryUUID.java [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+
+    for (int i = 0; i < 8; i++) msb = (msb << 8) | (randomBytes[i] & 0xff);
+
+    for (int i = 8; i < 16; i++) lsb = (lsb << 8) | (randomBytes[i] & 0xff);
+
+    return new UUID(msb, lsb);
+  }
+
+  private static class Holder {
+    static final Random numberGenerator = new Random();
```
Comment (adinauer) [inline sentry/src/main/java/io/sentry/SentryUUID.java:295]: According to https://docs.oracle.com/javase/7/docs/api/java/util/Random.html there might be poor performance when using `Random` from multiple threads.

> Instances of java.util.Random are threadsafe. However, the concurrent use of the same java.util.Random instance across threads may encounter contention and consequent poor performance. Consider instead using [ThreadLocalRandom](https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/ThreadLocalRandom.html) in multithreaded designs.

Should we vendor `ThreadLocalRandom` as well and use that?

Did you also do (performance) testing in a multi threaded environment?

File: sentry/src/main/java/io/sentry/SentryUUID.java (hunk 3 above)
Comment (adinauer) [reply sentry/src/main/java/io/sentry/SentryUUID.java:70]: I presume no, as long as we use String values in the SDK.

File: sentry/src/main/java/io/sentry/SentryUUID.java (hunk 5 above)
Comment (lbloder) [reply sentry/src/main/java/io/sentry/SentryUUID.java:295]: Hmm, good point. No, didn't do multithreaded perf testing. could try to spin up something to compare `Random` to `ThreadLocalRandom`

File: sentry/src/main/java/io/sentry/SentryUUID.java [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+  }
+
+  /**
+   * Returns a string representation of the given UUID. The returned string is formatted as
+   * described in {@link UUID#toString()}.
+   *
+   * @param uuid the UUID to represent as a string
+   * @return a string representation of the given UUID
+   */
+  public static String toString(final UUID uuid) {
```
[the remark this answers is not in the document]
Comment (lbloder) [reply sentry/src/main/java/io/sentry/SentryUUID.java:135]: Great, will remove this method then.

File: sentry/src/main/java/io/sentry/SentryUUID.java (hunk 5 above)
Comment (lbloder) [reply sentry/src/main/java/io/sentry/SentryUUID.java:295]: Yes, this does make quite a difference. Below is a comparison for UUID generation:

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/lukasbloder/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/lukasbloder/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">

</head>

<body link="#467886" vlink="#96607D">


Benchmark | Score |   | Error | Units | Threads | Comparison
-- | -- | -- | -- | -- | -- | --
uuidGenerationRandom | 23789834,814 | ± | 117697,18 | ops/s | 1 | 100%
uuidGenerationRandomMultiThreaded | 2280766,845 | ± | 263522,36 | ops/s | 10 | 10%
uuidGenerationThreadLocalRandomMultiThreaded | 185639075,638 | ± | 5090316,9 | ops/s | 10 | 780%



</body>

</html>

Comment (adinauer) [reply sentry/src/main/java/io/sentry/SentryUUID.java:295]: whew, what's the baseline we're comparing to here? are all of those using the FastUUID code just with a different `Random` implementation?

Comment (lbloder) [reply sentry/src/main/java/io/sentry/SentryUUID.java:295]: The code from `FastUUID` is only responsible for converting the `UUID` to string. This was not done in this test.

I tested the `SentryUUID.randomUUID()` function, which is a clone of the `java.util.UUID.randomUUID()` function with `java.util.Random` as the random generator.

So, Results 1 and 2 test the above mentioned function in single-threaded vs multi-threaded scenarios.

Result 3 tests the same function with `java.util.Random` replaced with `java.util.concurrent.ThreadLocalRandom`. I replaced `Random random = SentryUUID.Holder.numberGenerator;` with `Random random = ThreadLocalRandom.current();`

Comment (lbloder) [reply sentry/src/main/java/io/sentry/SentryUUID.java:295]: Also, the benchmark code can now be found here: https://github.com/lbloder/senty-uuid-bench

**Review by adinauer — APPROVED** [review]

**Comment by lbloder** [comment]
Did some additional Testing for MultiThreading based on [this](https://github.com/getsentry/sentry-java/pull/3818#discussion_r1820211349) comment from @adinauer.

The results below show Test runs for:
1) `java.util.Random` single-threaded
2) `java.util.random` multi-threaded with 10 Threads
3) `java.util.concurrent.ThreadLocalRandom` with 10 Threads
4) A version in which we manage a `ThreadLocal<java.util.Random>` ourselves

## Results
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/lukasbloder/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/lukasbloder/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">

</head>

<body link="#467886" vlink="#96607D">


Benchmark | Score |   | Error | Units | Threads | Comparison |  
-- | -- | -- | -- | -- | -- | -- | --
uuidGenerationRandom | 23829634,539 | ± | 101112,796 | ops/s | 1 | 100% |  
uuidGenerationRandomMultiThreaded | 2458588,180 | ± | 339397,214 | ops/s | 10 | 10% |  
uuidGenerationThreadLocalRandomMultiThreaded | 180884073,702 | ± | 2482343,414 | ops/s | 10 | 759% | 157%
uuidGenerationThreadLocalRandomMultiThreadedSelfManaged | 114937169,398 | ± | 1555655,255 | ops/s | 10 | 482% | 100%



</body>

</html>

We can see, that `ThreadLocalRandom` is highly optimized, outpacing our self-managed version by ~50%. However, we still get an almost 5x increase compared to using a single `Random` instance.

## Secure Random
Out of curiosity, I also did a benchmark run to compare `UUID.randomUUID` in single-threaded and multi-threaded scenarios:

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/lukasbloder/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/lukasbloder/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">

</head>

<body link="#467886" vlink="#96607D">


Benchmark | Score |   | Error | Units | Threads | Comparison
-- | -- | -- | -- | -- | -- | --
uuidGenerationSecureRandom | 3229714,852 | ± | 13916,035 | ops/s | 1 | 100%
uuidGenerationSecureRandomMultiThreaded | 1567325,134 | ± | 36688,679 | ops/s | 10 | 49%



</body>

</html>

We can see, that SecureRandom performance is also impacted when running in multiple threads, however, not nearly as much as Random.

**Pull request merged**